src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -391,6 +391,10 @@ private void run() throws IOException {
         if (doneSomeWork) {
             saveProgram();
         }
+        if (threads.isEmpty()) {
+            execution.cleanUpHeap();
+            scripts.clear();
```

Comment:
Clearing `scripts` was also necessary, or even with the fix to the timing of `SerializableClassRegistry`, the test would still fail on Jenkins 1. The failure showed no root references, so I had to guess at the cause.